skimage/morphology/rolling_ball.py

```diff
@@ -0,0 +1,148 @@
+    >>> from skimage.morphology import rolling_ball
+
+    Subtract a black background from an example image
+
+    >>> image = data.coins()
+    >>> result, bg = rolling_ball(image, radius=100)
+
+    Subtract a white background from an example image
+
+    >>> image = data.page()
+    >>> result, bg = rolling_ball(image, radius=100, white_background=True)
+
+    Notes
+    -----
+    Check out the gallery example of this function (TODO: link)
+
+    References
+    ----------
+    .. [1] Sternberg, Stanley R. "Biomedical image processing." Computer 1
+           (1983): 22-34. :DOI:`10.1109/MC.1983.1654163`
+    """
+
+    if not isinstance(radius, float):
+        try:
+            radius = float(radius)
+        except ValueError:
+            raise ValueError(f"Radius should be float, was {type(radius)}")
+
+    if radius <= 0:
+        raise ValueError(f"Radius must be greater zeros, was {radius}")
+
+    try:
+        white_background = bool(white_background)
+    except ValueError:
+        raise ValueError(
+            f"white_background must be convertible to a boolean, "
+            f"was {type(white_background)}"
+        )
+
+    if not isinstance(input_img, np.ndarray):
+        raise ValueError(
+            f"input_img must be a np.ndarray, "
+            f"was {type(input_img)}"
+        )
+
+    if not issubclass(input_img.dtype.type, np.integer):
+        raise ValueError("Currently only integer images are supported.")
+
+    img = input_img.copy()
+    if white_background:
+        img = invert(img)
+
+    # sagitta assuming the position is where the ball touches
+    # the image umbra
+    L = np.arange(-radius, radius + 1)
+    X, Y = np.meshgrid(L, L)
+    distance = np.sqrt(X ** 2 + Y ** 2)
+    sagitta = radius - np.sqrt(radius ** 2 - distance ** 2)
+    sagitta[np.isnan(sagitta)] = np.Inf
+
+    kernel = np.array((X ** 2 + Y ** 2) <= radius ** 2, dtype=float)
+    kernel[kernel == 0] = np.Inf
+
+    kernel_size = np.array(kernel.shape)
+    img_original = img.copy()
+    img_size = img.shape
+    x = img
+    stride = (1, 1)
+
+    # pad the image
+    padding_amount = kernel_size // 2 * 2
+    half_pad = padding_amount // 2
+    img = np.Inf * np.ones(x.shape + padding_amount, dtype=x.dtype)
+    img[half_pad[0]:-half_pad[0], half_pad[1]:-half_pad[1]] = x
+    large_img_size = img.shape
+
+    # indices of kernel in image coords
+    x_idx = np.arange(kernel_size[1])
+    y_idx = large_img_size[1] * np.arange(kernel_size[0])
+    kernel_idx = (x_idx[np.newaxis, :] + y_idx[:, np.newaxis]).flatten()
+
+    # indices corresponding to each ancor of the kernel
+    # (top left corner instead of center)
+    x_ancors = np.arange(
+        large_img_size[1] - kernel_size[1] + 1, step=stride[1])
+    y_ancors = large_img_size[1] * \
+        np.arange(large_img_size[0] - kernel_size[0] + 1, step=stride[0])
+    ancor_offsets = (x_ancors[np.newaxis, :] +
+                     y_ancors[:, np.newaxis]).flatten()
```

Comment:
This seems to be the meat of the algorithm, but I don't quite understand what you are building... Is this a set of linear offsets from a central coordinate? Can you provide a tiny numerical example for illustration?

Reply:
also, ancor -> anchor. =)

Reply:
fixed the ancor typo (why is typing so hard)?

Reply:
actually, the meat of the algorithm is the block following the ones you referenced.

The big picture is that this piece of code (re-)implements 2d convolution, but with a twist to enable the rolling ball algorithm. Here, I'm implementing the version that interprets convolution as matrix multiplications. ([This blog post](https://sgugger.github.io/convolution-in-depth.html) provides a nice overview of that idea.)

For this, we need to identify the indices of pixels affected by a single application of the kernel. Given an anchor for the kernel (top-left corner), the first block of code computes those indices relative to the anchor. The second block of code then computes all anchor points where the kernel must be applied.

The third block (that you didn't reference) combines each anchor with the relative window coordinates (`filter_idx = anchor_offsets[...`), extracts the affected image intensities, and applies the kernel (by multiplying with it). At this point, there are 3 major differences from standard convolution to make the rolling ball method work (which prevent us from using an existing conv2d implementation):

1. Instead of just doing `px_value * kernel` we have to add the value of the sagitta to each intensity value first `(px_value + sagitta) * kernel`. This is the only difference between `rolling_ball` and the `rolling_disk` approximation.
2. We don't reduce the vector using a sum (like in standard convolution), but, instead, use min.
3. The kernel is large (`>>(50,50)`), so we can't fit the entire `px_value` matrix into memory; instead we have to do it in batches.

Does this shed some light on what's going on?

Reply:
Thanks @FirefoxMetzger! That's very helpful! Still a bit fuzzy, but much less so than before. Some follow-up comments/questions:

> I'm implementing the version that interprets convolution as matrix multiplications

Ah! I know exactly what you mean, and in fact we use this trick in our [filters tutorial](https://github.com/jni/skimage-tutorials/blob/monash-df2020-05-completed/lectures/1_image_filters.ipynb)! (Search for "the simplest neural network".)

Given that code, I wonder if using `view_as_windows` would make this code simpler:

```python
windowed = util.view_as_windows(padded_image, kernel_shape)
for coords_slice in iter_chunks(windowed, chunk_size):  # <-- needs to be implemented
    matrix = windowed[coords_slice].reshape((-1, kernel_size))
    filtered = np.min((matrix + sagitta) * kernel, axis=1)
    output[coords_slice] = filtered
```

? (There might need to be some futzing around with the slices given padding/cropping, but that's the basic idea I think.)

BUT,

Another thought: should we instead be using [`scipy.ndimage.generic_filter`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.ndimage.generic_filter.html), together with a [low-level callable](https://docs.scipy.org/doc/scipy/reference/generated/scipy.LowLevelCallable.html#scipy.LowLevelCallable)?

Reply:
`util.view_as_window` seems quite useful indeed. We might be able to remove some of the current index magic. The problem is memory constraints because `windowed.shape == (img.size, kernel.size)` and we are dealing with excessively large kernels (`>>(50,50)`). I will see if I can intelligently slice the input image and then use `util.view_as_window` inside the loop.

I didn't know about `scipy.ndimage.generic_filter`. I don't see the generalization (yet?) though. My problem is that filters usually compute the sum (whereas we need min). Another struggle is the constant (sagitta) that we add to each window. If I am not mistaken, then `generic_filter` will apply the kernel to the image and not allow to add such a constant; I might not properly understand the function though.

Reply:
> I will see if I can intelligently slice the input image and then use util.view_as_window inside the loop.

Note that `view_as_windows` creates a view, not a copy, using clever NumPy stride tricks. You *can* and should create the windows outside the loop — they are free. Only the reshape actually triggers a copy and costs memory!

If you are not familiar with NumPy stride tricks, you can check out these notebooks for details on how this works:

https://github.com/jni/scipy-japan-2019

It is a Damn Cool feature of NumPy. =D

> I didn't know about scipy.ndimage.generic_filter. I don't see the generalization (yet?) though.

Generic filter lets you compute whatever reduction you want. Your input is a set of values (the stuff under the footprint) and the output is a single value (the result). What happens in the middle is up to you. See this little blog post I wrote a while back, where I use (drumroll) a minimum filter as the example. 😂 

https://ilovesymposia.com/2017/03/12/scipys-new-lowlevelcallable-is-a-game-changer/

You should be able to pass an arbitrary number of kernel weights to the function, but I have never successfully done this with a Numba/Cython function, because it ends up in a `void *user_data` argument. It would 100% be worth exploring though! It would provide both a proof of principle documentation for that feature, and reduce this function to constructing some kernels and then calling `generic_filter`.

Reply:
Note: I just found some examples of using `user_data` with Cython here:

https://docs.scipy.org/doc/scipy/reference/tutorial/ndimage.html#ndimage-ccallbacks

I think the "easiest"/least hard path would be to create an array that contains both the kernel and the sagitta and then grab those individually within Cython, since Cython understands NumPy arrays. This might be harder than the indexing in the first instance, though, but it would probably be much faster.

Reply:
I refactored the code to use `view_as_window`. I didn't know numpy stride tricks were a think. They are quite nifty indeed. It makes the code a lot shorter.

I will look into `generic_filter` and see if I can get a better understanding of how it works and how to apply it here.

Reply:
Okay, it can indeed be easily implemented with generic_filter:

```
def fn(px_values):
    return np.min((px_values + flat_sagitta) * flat_kernel, axis=-1)

background = generic_filter(
    img, fn, size=kernel.shape, mode='constant', cval=np.iinfo(image.dtype).max)
```

The problem I encountered here is that this is again very memory hungry. I will check if using `lowlevelcallable` can improve this.

Reply:
Yeah tbh I knew about the memory issues but didn't know if the kernels were *quite* big enough to hit them:

https://github.com/scipy/scipy/issues/4878

LLC makes things faster, but no less memory-hungry.

So `view_as_windows` it is! Thanks for exploring with me!

Reply:
Haha, it's an interesting problem so the pleasure is mine. That said, I think this can still be made fast and memory-efficient with cython (though, I have never written anything serious with cython). 

The memory problem comes from constructing an intermediate array (`chunk` in firefoxmetzger/scikit-image@730d809), but this only really happens because we have to go back up to the python layer only to pass the reference back down to numpy when we call the filter. With cython it should be possible to avoid constructing the temporary array and instead lookup the values and only store the result. No idea if this will play along with hardware acceleration though.

Reply:
"fixed" as of firefoxmetzger/scikit-image@e80a4f3. I added a cython function to apply the kernel that is very memory efficient, and (perhaps as a result) surprisingly fast. Now the question is how much my naive cython code can be optimized. I think I saw at least one point where parts can be replaced with a BLAS call, but I have to figure out how that works with cython first.

I will mark this discussion as resolved.